### PR TITLE
Added namespace configurator for existing user SSH keys

### PR DESCRIPTION
**Pull request by mshaposhnik**
Signed-off-by: Max Shaposhnik <mshaposh@redhat.com>

### What does this PR do?
Added namespace configurator for mounting existing user SSH keys into namespace during provision

### Screenshot/screencast of this PR
Generated secret data will looks like:

image

### What issues does this PR fix or reference?
[](https://github.com/eclipse/che/issues/20832)
https://github.com/eclipse/che/issues/20832

### How to test this PR?

 - Deploy Che with devworkspaces enabled;
 - Add SSH key via Swagger API (since SSH plugin in Theia is disabled on DVW mode)
 - Remove user namespace and refresh dashboard to force namespace provision again
 - Check secret is created

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+      throws InfrastructureException {
+    List<SshPairImpl> sshPairs;
+    try {
+      sshPairs = sshManager.getPairs(context.getUserId(), "vcs");
+    } catch (ServerException e) {
+      String message = format("Unable to get SSH Keys. Cause: %s", e.getMessage());
+      LOG.warn(message);
+      throw new InfrastructureException(e);
+    }
+    //    Q: do we need that generation for devworkspaces ?
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java:119]: I'm not sure here -- what would this be used for?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            .withType("generic")
+            .withMetadata(buildMetadata())
+            .build();
+
+    client.secrets().inNamespace(namespaceName).createOrReplace(secret);
+  }
+
+  private ObjectMeta buildMetadata() {
+    return new ObjectMetaBuilder()
+        .withName("git-ssh-key")
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java:172]: Should this be a constant somewhere? Also, should we prefix with `che-`?

**Review by amisevsk — APPROVED** [review]
LGTM from a DWO perspective.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java (hunk 1 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java]: I do not remember the reason why it was added in Che7. Let's remove it for now :)

**Comment by skabashnyuk** [comment]
@amisevsk @JPinkney can you confirm that the path of ssh key should be `/.ssh/` as described here https://github.com/devfile/devworkspace-operator/issues/613#issuecomment-978202198

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    this.sshManager = sshManager;
+    this.clientFactory = clientFactory;
+  }
+
+  @Override
+  public void configure(NamespaceResolutionContext namespaceResolutionContext, String namespaceName)
+      throws InfrastructureException {
+
+    var client = clientFactory.create();
+    List<SshPairImpl> vcsSshPairs = getVcsSshPairs(namespaceResolutionContext);
```
Comment (ibuziuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java:78]: @amisevsk wondering if we need to process the internal ssh keys for async-storage. I'm not sure how the DevWorkspace handles the key management for it

Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/SshKeysConfigurator.java:78]: https://github.com/eclipse-che/che-server/blob/98b66ff1f64ebf2174b51e1a348d482590962e82/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStorageProvisioner.java#L204

**Comment by mshaposhnik** [comment]
Yes, correct path should be `/etc/ssh/ssh_config`, pushed the fix.

**Comment by amisevsk** [comment]
`/etc/ssh` should work, however note that currently async storage mounts its own ssh key to `/etc/ssh/private` for the async storage sidecar, so that's a sticking point we may need to work around in some way.

**Comment by mshaposhnik** [comment]
Maybe separate secret with different mount path-s for them ? (but ssh_config file should remain the same IMHO....)

**Comment by amisevsk** [comment]
> Maybe separate secret with different mount path-s for them ? (but ssh_config file should remain the same IMHO....)

I think DWO needs to isolate the async-storage sidecar from automount, etc. It only has a single responsibility (communicate with async server) and so should not be impacted by anything else. I've created https://github.com/devfile/devworkspace-operator/issues/707 to track this issue.

**Comment by skabashnyuk** [comment]
[crw-ci-test --rebuild]

**Comment by skabashnyuk** [comment]
@amisevsk @ibuziuk does it make sense to handle async-storage keys as a separate task? Because it is not really clear what to do now.

**Comment by mshaposhnik** [comment]
@skabashnyuk Yes that's why https://github.com/devfile/devworkspace-operator/issues/707 created AFAIU

**Comment by amisevsk** [comment]
To clarify, DWO currently creates the keypair for the async storage usecase, so no intervention from Che Server is needed to make that happen (Che should not have to do anything for the async storage case). The issue is that currently the keypair provisioned by DWO will collide with an automount to that directory, which is what's described in devfile/devworkspace-operator#707.

**Pull request merged**